climada_petals/hazard/tc_surge_bathtub.py

```diff
@@ -97,11 +96,15 @@ def from_tc_winds(wind_haz, topo_path, inland_decay_rate=0.2, add_sea_level_rise
         # Elevation data and coastal/non-coastal indices are used later in the code
         coastal_centroids_h = coastal_centroids_h[elevation_msk]
         coastal_idx = coastal_msk.nonzero()[0]
-        noncoastal_idx = (~coastal_msk).nonzero()[0]
+
+        # to each centroid, assign its position in `coastal_idx`, and assign an out-of-bounds
+        # index to all centroids that are not contained in `coastal_idx`
+        cent_to_coastal_idx = np.full(coastal_msk.shape, coastal_idx.size, dtype=np.int64)
+        cent_to_coastal_idx[coastal_msk] = np.arange(coastal_idx.size)
 
         # Initialize intensity array at coastal centroids
         inten_surge = wind_haz.intensity.copy()
-        inten_surge[:,noncoastal_idx] *= 0
+        inten_surge.data[~coastal_msk[inten_surge.indices]] = 0
```

Comment:
The previous code would create a huge dense array in memory. In extreme cases, this array would be so huge that indexing operations on it would fail with a very cryptic error message (`ValueError: could not convert integer scalar`), or the process would be killed due to out-of-memory.

The new code looks complicated, but it's really preventing that any temporary arrays are created that are larger than the amount of non-zero data stored in the sparse matrix.